xtask/src/diff_walk.rs

```diff
@@ -0,0 +1,182 @@
+        FileContent::Dir
+    } else {
+        let data = fs.read(&path)?;
+        let hash = format!("{:x}", Sha256::digest(data));
+        FileContent::Regular(hash)
+    };
+    Ok(WalkDirEntry {
+        path: dir_entry.path().into(),
+        content,
+        mode: metadata.mode(),
+    })
+}
+
+fn walk_with_lib(
+    fs: &Ext4,
+    path: ext4_view::Path<'_>,
+) -> Result<Vec<WalkDirEntry>> {
+    let mut output = Vec::new();
+
+    output.push(WalkDirEntry {
+        path: ext4_view::PathBuf::from(path).into(),
+        content: FileContent::Dir,
+        mode: fs.symlink_metadata(path)?.mode(),
+    });
+
+    for entry in fs.read_dir(path)? {
+        let entry = entry?;
+        let path = entry.path();
+        let name = entry.file_name();
+        if name == "." || name == ".." {
+            continue;
+        }
+
+        // TODO: use DirEntry::file_type once that exists.
+        if fs.symlink_metadata(&path)?.is_dir() {
+            output.extend(walk_with_lib(fs, path.as_path())?);
+        } else {
+            output.push(new_dir_entry(fs, entry)?);
+        }
+    }
+
+    Ok(output)
+}
+
+/// Check that walking the filesystem with the `ext4-view` crate gives
+/// the same results as mounting the filesystem and walking it with
+/// [`std::fs`].
+///
+/// See `./bin/mount_and_walk.rs` for details of mounting and walking
+/// the filesystem. That program is run under `sudo` since `mount`
+/// requires elevated permissions.
+pub fn diff_walk(path: &Path) -> Result<()> {
+    // Build `mount_and_walk` in release mode.
+    let status = Command::new("cargo")
+        .args([
+            "build",
+            "--release",
+            "--package",
+            "xtask",
+            "--bin",
+            "mount_and_walk",
+        ])
+        .status()?;
+    if !status.success() {
+        bail!("failed to build mount_and_walk");
+    }
+
+    let actual = {
+        let ext4 = Ext4::load_from_path(path)?;
+        let mut paths = walk_with_lib(&ext4, ext4_view::Path::ROOT)?;
+        paths.sort_unstable();
+        paths
+            .iter()
+            .map(|dir_entry| dir_entry.format())
+            .collect::<Vec<_>>()
+    };
+    let expected = {
+        let output = Command::new("sudo")
+            .arg("target/release/mount_and_walk")
+            .arg(path)
+            .output()?;
+        assert!(output.status.success());
+        let mut lines = output
+            .stdout
+            .split(|c| *c == b'\n')
+            .map(|l| l.to_vec())
+            .collect::<Vec<_>>();
+
+        // Remove the empty line at the end of the file.
+        let last = lines.pop().unwrap();
+        assert!(last.is_empty());
+
+        lines
+    };
+
+    for (a, b) in actual.iter().zip(expected.iter()) {
+        if a != b {
+            println!(
+                "{} != {}",
+                String::from_utf8_lossy(a),
+                String::from_utf8_lossy(b)
+            );
+            process::exit(1);
```

Comment:
why asserts elsewhere but `process::exit` here?

Reply:
Seems I was accidentally mixing several error styles (assert, exit, and Result). I've replaced the asserts with error returns via `bail!`. For mismatches between the actual and expected output I'm using println + exit because those are "expected" errors, so I want the error to be simple and readable (no extra clutter such as backtrace). For an unexpected error such malformed output I don't care as much how the error looks.